### ISSUE TITLE
true semantics of "/$pat/o"

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8344]
Migrated from [rt.perl.org#38625](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=38625) (status was 'rejected')

Searchable as RT38625$

**Comment by p5pRT** [comment]
### From Ulrich.Windl@rz.uni-regensburg.de

#### Created by Ulrich.Windl@rz.uni-regensburg.de


The exact semantics of /$pattern/o is not clear&#8203;: The docs say the pattern
is compiled only once\. It may seem clear if considering static blocks of
code\. However&#8203;:

\{
&nbsp;    my foo = "abc";

&nbsp;    \.\.\. /$foo/o
\}

\{
&nbsp;    my foo = "cde";
&nbsp;    \.\.\. /$foo/o
\}

Naturally one would expect that in different blocks each $foo pattern is
compiled once \(twice alltogether in this example\)\.

Now consider this&#8203;:

sub make\_matcher\($\)
\{
&nbsp;    my $epattern = shift;

&nbsp;    if \($epattern =~ /^\!\./\) \{
&nbsp;        $epattern = substr\($epattern\, 1\);

&nbsp;        return sub \{
&nbsp;            my $arg = shift;

&nbsp;            return $arg \!~ /$epattern/o;
&nbsp;        \};
&nbsp;    \} else \{
&nbsp;        return sub \{
&nbsp;            my $arg = shift;

&nbsp;            return $arg =~ /$epattern/o;
&nbsp;        \};
&nbsp;    \}
\}

my $matcher = make\_matcher\("abc"\)
\.\.\. &$matcher\("test"\)

my $matcher = make\_matcher\("\!cde"\)
\.\.\. &$matcher\("test"\)

Isn't $epattern a different variable for each call of make\_matcher\(\)?
Thus it should be compiled to a new pattern \(when creating a sub\)\, but not
when the returned sub is called\.

However all subsequent matchers use the first match pattern\. I thinks that's
a bug\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.3:

Configured by ActiveState at Thu May 27 17:36:19 PDT 2004.

Summary of my perl5 (revision 5.0 version 8 subversion 3) configuration:
  Platform:
    osname=hpux, osvers=11.22, archname=IA64.ARCHREV_0-thread-multi
    uname='hp-ux bigsur03 b.11.22 u ia64 1800516905 unlimited-user license '
    config_args='-ders -Dcc=cc -Accflags=-DNO_HASH_SEED -Dusethreads -Duseithreads 
-Ud_sigsetjmp -Uinstallusrbinperl -Ulocincpth= -Uloclibpth= -Dd_attribut=undef -
Dd_u32align=define -Doptimize=-fast +DSitanium2 +Ofltacc=strict -Duselargefiles -
Dinc_version_list=5.8.2/$archname 5.8.2 5.8.1/$archname 5.8.1 5.8.0/$archname 
5.8.0 -Duseshrplib -Dprefix=/opt/perl -Dcf_by=ActiveState -
Dcf_email=support@ActiveState.com'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define 
usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags =' -D_POSIX_C_SOURCE=199506L -D_REENTRANT -Ae -D_HPUX_SOURCE 
-Wl,+vnocompatwarnings -DNO_HASH_SEED -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 
',
    optimize='-fast +DSitanium2 +Ofltacc=strict',
    cppflags='-D__STDC_EXT__ -D_HPUX_SOURCE -D_POSIX_C_SOURCE=199506L -D_REENTRANT 
-D_HPUX_SOURCE -Wl,+vnocompatwarnings -DNO_HASH_SEED'
    ccversion='', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='/usr/bin/ld', ldflags =' -L/usr/lib/hpux32'
    libpth=/usr/lib/hpux32 /lib /usr/lib /usr/ccs/lib /usr/local/lib
    libs=-lnsl -lnm -lndbm -ldl -ldld -lm -lsec -lpthread -lc
    perllibs=-lnsl -lnm -ldl -ldld -lm -lsec -lpthread -lc
    libc=/usr/lib/hpux32/libc.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_hpux.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-B,deferred 
'
    cccdlflags='+Z', lddlflags='-b +vnocompatwarnings -L/usr/lib/hpux32'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY
    22218 Remove the caveat about detached threads crashing on Windows
    22201 Avoid threads+win32 crash by freeing Perl interpreter slightly later
    22169 Display 'out of memeory' errors using low-level I/O
    22159 Upgrade to Time::Hires 1.55
    22120 Make 'Configure -Dcf_by=...' work
    22051 Upgrade to Time::HiRes 1.54
    21540 Fix backward-compatibility issues in if.pm


@INC for perl v5.8.3:
    /opt/perl/lib/5.8.3/IA64.ARCHREV_0-thread-multi
    /opt/perl/lib/5.8.3
    /opt/perl/lib/site_perl/5.8.3/IA64.ARCHREV_0-thread-multi
    /opt/perl/lib/site_perl/5.8.3
    /opt/perl/lib/site_perl
    .


Environment for perl v5.8.3:
    HOME=/home/windl
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    
PATH=/usr/local/bin:/usr/local/bin:/usr/bin:/usr/ccs/bin:/usr/contrib/bin:/usr/con
trib/Q4/bin:/opt/perl/bin:/opt/ipf/bin:/opt/hparray/bin:/opt/nettladm/bin:/opt/fcm
s/bin:/opt/sec_mgmt/bastille/bin:/opt/resmon/bin:/opt/gnome/bin:/usr/bin/X11:/opt/
ignite/bin:/usr/contrib/kwdb/bin:/opt/wbem/bin:/opt/wbem/sbin:/opt/graphics/common
/bin:/opt/sec_mgmt/spc/bin:/opt/java1.4/jre/bin:/opt/spb/bin:/opt/hpsmh/bin:/opt/u
pgrade/bin:/opt/gwlm/bin:/opt/mozilla:/usr/contrib/bin/X11:/opt/ssh/bin
    PERL_BADLANG (unset)
    SHELL=/usr/bin/csh



```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, Feb 24\, 2006 at 04&#8203;:29&#8203;:02AM \-0800\, Ulrich Windl wrote&#8203;:
> Isn't $epattern a different variable for each call of make\_matcher\(\)?
> Thus it should be compiled to a new pattern \(when creating a sub\)\, but not
> when the returned sub is called\.
> 
> However all subsequent matchers use the first match pattern\. I thinks that's
> a bug\.

perlop says that /o means "compile pattern only once"\. Which is exactly
what happens\. The fact that multiple closures share the same pattern
doesn't alter that fact that the pattern will only be compiled once\.

\-\- 
Monto Blanco\.\.\. scorchio\!

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @ysth

On Fri\, Feb 24\, 2006 at 04&#8203;:29&#8203;:02AM \-0800\, Ulrich Windl wrote&#8203;:
> The exact semantics of /$pattern/o is not clear&#8203;: The docs say the pattern
> is compiled only once\. It may seem clear if considering static blocks of
> code\. However&#8203;:
> 
> \{
>     my foo = "abc";
> 
>     \.\.\. /$foo/o
> \}
> 
> \{
>     my foo = "cde";
>     \.\.\. /$foo/o
> \}
> 
> Naturally one would expect that in different blocks each $foo pattern is
> compiled once \(twice alltogether in this example\)\.

It's not having different variables that does that\, it's having multiple
patterns in your source\.

This may compile twice\, too&#8203;:

&nbsp;    my $foo = "abc";
&nbsp;    \.\.\. /$foo/o;
&nbsp;    \.\.\. /$foo/o;

whether or not $foo has the same value\.  But this only compiles once&#8203;:

&nbsp;    my $foo = "abc";
&nbsp;    for \(1\, 2\) \{
&nbsp;      \.\.\. /$foo/o;
&nbsp;    \}

> Now consider this&#8203;:
> 
> sub make\_matcher\($\)
> \{
>     my $epattern = shift;
> 
>     if \($epattern =~ /^\!\./\) \{
>         $epattern = substr\($epattern\, 1\);
> 
>         return sub \{
>             my $arg = shift;
> 
>             return $arg \!~ /$epattern/o;
>         \};
>     \} else \{
>         return sub \{
>             my $arg = shift;
> 
>             return $arg =~ /$epattern/o;
>         \};
>     \}
> \}
> 
> my $matcher = make\_matcher\("abc"\)
> \.\.\. &$matcher\("test"\)
> 
> my $matcher = make\_matcher\("\!cde"\)
> \.\.\. &$matcher\("test"\)
> 
> Isn't $epattern a different variable for each call of make\_matcher\(\)?

Doesn't matter\.

> Thus it should be compiled to a new pattern \(when creating a sub\)\, but not
> when the returned sub is called\.

The compilation does have to happen during the sub call\, since it
is supposed to use the value the interpoloated variable has then\, not
its value when the sub\{\} is returned\.

> However all subsequent matchers use the first match pattern\. I thinks that's
> a bug\.

I can see it working either way\, in the case of an //o in an anonymous
sub\.

Obviously removing the //o works around your problem\.  Or you could&#8203;:

sub make\_matcher\($\)
\{
&nbsp;    my $epattern = shift;

&nbsp;    if \($epattern =~ /^\!\./\) \{
&nbsp;        $epattern = substr\($epattern\, 1\);
&nbsp;        $epattern = qr/$epattern/;

&nbsp;        return sub \{
&nbsp;            my $arg = shift;

&nbsp;            return $arg \!~ $epattern;
&nbsp;        \};
&nbsp;    \} else \{
&nbsp;        $epattern = qr/$epattern/;
&nbsp;        return sub \{
&nbsp;            my $arg = shift;

&nbsp;            return $arg =~ $epattern;
&nbsp;        \};
&nbsp;    \}
\}

**Comment by p5pRT** [comment]
### From @ysth

On Fri\, Feb 24\, 2006 at 05&#8203;:25&#8203;:38PM \+0000\, Dave Mitchell wrote&#8203;:
> On Fri\, Feb 24\, 2006 at 04&#8203;:29&#8203;:02AM \-0800\, Ulrich Windl wrote&#8203;:
> > Isn't $epattern a different variable for each call of make\_matcher\(\)?
> > Thus it should be compiled to a new pattern \(when creating a sub\)\, but not
> > when the returned sub is called\.
> > 
> > However all subsequent matchers use the first match pattern\. I thinks that's
> > a bug\.
> 
> perlop says that /o means "compile pattern only once"\. Which is exactly
> what happens\. The fact that multiple closures share the same pattern
> doesn't alter that fact that the pattern will only be compiled once\.

Would you expect different threads to also "compile pattern only once"?
\(I'm not going to try\.\)

Isn't there a parallel from the threads case to the multiple anonymous
subs case?  \(I think closure or non\-closure isn't relevant\.\)

**Comment by p5pRT** [comment]
### From Ulrich.Windl@rz.uni-regensburg.de

I Am still missing of a precise definition what "the pattern" that is only 
compiled once is&#8203;: It seems to be the "code location \(location of pattern in code\)" 
rather than "the pattern"\.

Regards\,
Ulrich

**Comment by p5pRT** [comment]
### From @hvds

"Ulrich Windl" \<ulrich\.windl@&#8203;rz\.uni\-regensburg\.de> wrote&#8203;:
:I Am still missing of a precise definition what "the pattern" that is only 
:compiled once is&#8203;: It seems to be the "code location \(location of pattern in code\)" 
:rather than "the pattern"\.

Yes&#8203;: the first time this code is executed\, the pattern is compiled\. Any
subsequent time the code is executed\, the pattern is not recompiled\.

Precisely how this interacts I'm not sure&#8203;: I believe the flag is stored
in the optree\, so I imagine anything that creates a second chunk of code
by cloning the optree \(eg spawning a thread?\) before the pattern has been
compiled will give you two independent patterns each of which will be
compiled the first time it is executed\.

Hugo

**Comment by p5pRT** [comment]
### From @rgs

hv@&#8203;crypt\.org wrote&#8203;:
> Precisely how this interacts I'm not sure&#8203;: I believe the flag is stored
> in the optree\, so I imagine anything that creates a second chunk of code
> by cloning the optree \(eg spawning a thread?\) before the pattern has been
> compiled will give you two independent patterns each of which will be
> compiled the first time it is executed\.

Yes\, that's the flag PMf\_KEEP in the optree\.
But\, IIRC\, the optree is never cloned\, so it will actually be compiled
only once\. \(From what I see in pp\_regcomp\(\) this optree modification
shouldn't cause thread safety problems\)\.

**Comment by p5pRT** [comment]
### From @tamias

On Mon\, Feb 27\, 2006 at 08&#8203;:24&#8203;:27AM \+0100\, Ulrich Windl wrote&#8203;:
> I Am still missing of a precise definition what "the pattern" that is
> only compiled once is&#8203;: It seems to be the "code location \(location of
> pattern in code\)" rather than "the pattern"\.

What definition are you using for "the pattern"?

Ronald

**Comment by p5pRT** [comment]
### From Ulrich.Windl@rz.uni-regensburg.de

On 27 Feb 2006 at 10&#8203;:08\, Ronald J Kimball wrote&#8203;:

> On Mon\, Feb 27\, 2006 at 08&#8203;:24&#8203;:27AM \+0100\, Ulrich Windl wrote&#8203;:
> > I Am still missing of a precise definition what "the pattern" that is
> > only compiled once is&#8203;: It seems to be the "code location \(location of
> > pattern in code\)" rather than "the pattern"\.
> 
> What definition are you using for "the pattern"?

For me "the pattern" is the specific value seen the first time when the code in 
the containing block is evaluated \(not "compiled"\)\.
So when returning subroutines\, a new block of code is generated \(conceptually at 
least IMHO\)\, and therefore the pattern should be compiled with the current value 
in that new block\.

Regards\,
Ulrich

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'rejected'
